source/fe/fe_enriched.cc

```diff
@@ -0,0 +1,949 @@
+                                           const Quadrature< dim > &quadrature,
+                                           const Mapping< dim, spacedim > &mapping,
+                                           const typename Mapping< dim, spacedim >::InternalDataBase &mapping_internal,
+                                           const dealii::internal::FEValues::MappingRelatedData< dim, spacedim > &mapping_data,
+                                           const typename FiniteElement<dim,spacedim>::InternalDataBase &fe_internal,
+                                           internal::FEValues::FiniteElementRelatedData< dim, spacedim > &output_data
+                                          ) const
+{
+  const typename FE_Enriched<dim,spacedim>::InternalData
+  &fe_data = static_cast<const typename FE_Enriched<dim,spacedim>::InternalData &> (fe_internal);
+
+  // call FESystem's method to fill everything without enrichment function
+  fe_system.fill_fe_values(cell,
+                           cell_similarity,
+                           quadrature,
+                           mapping,
+                           mapping_internal,
+                           mapping_data,
+                           *fe_data.fesystem_data,
+                           output_data);
+
+  if (is_enriched)
+    {
+      multiply_by_enrichment(quadrature,
+                             fe_internal,
+                             mapping_data,
+                             cell,
+                             output_data);
+    }
+}
+
+template <int dim, int spacedim>
+const FESystem<dim,spacedim> &
+FE_Enriched<dim,spacedim>::
+get_fe_system() const
+{
+  return fe_system;
+}
+
+template <int dim, int spacedim>
+bool
+FE_Enriched<dim,spacedim>::
+hp_constraints_are_implemented () const
+{
+  return true;
+}
+
+template <int dim, int spacedim>
+void
+FE_Enriched<dim,spacedim>::
+get_face_interpolation_matrix (const FiniteElement<dim,spacedim> &source,
+                               FullMatrix<double>                &matrix) const
+{
+  if (const FE_Enriched<dim,spacedim> *fe_enr_other
+      = dynamic_cast<const FE_Enriched<dim,spacedim>*>(&source))
+    {
+      fe_system.get_face_interpolation_matrix(fe_enr_other->get_fe_system(),
+                                              matrix);
+    }
+  else
+    {
+      typedef FiniteElement<dim,spacedim> FEL;
+      AssertThrow(false,typename FEL::ExcInterpolationNotImplemented());
+    }
+}
+
+template <int dim, int spacedim>
+void
+FE_Enriched<dim,spacedim>::
+get_subface_interpolation_matrix (const FiniteElement<dim,spacedim> &source,
+                                  const unsigned int        subface,
+                                  FullMatrix<double>       &matrix) const
+{
+  if (const FE_Enriched<dim,spacedim> *fe_enr_other
+      = dynamic_cast<const FE_Enriched<dim,spacedim>*>(&source))
+    {
+      fe_system.get_subface_interpolation_matrix(fe_enr_other->get_fe_system(),
```

Comment:
ah, this is quite neat. I think I finally understand why you base everything on an fe system :-)